src/FastSerialization/StreamReaderWriter.cs

```diff
@@ -635,6 +636,19 @@ public override void Goto(StreamLabel label)
             uint offset = unchecked((uint)label - positionInStream);
             if (offset > (uint)endPosition)
             {
+                if(!inputStream.CanSeek)
+                {
+                    if((uint)label < positionInStream)
+                    {
+                        throw new Exception("Stream can not go backwards");
+                    }
+                    uint bytesToRead = offset - (uint)endPosition;
+                    for(int i = 0; i < bytesToRead; i++)
+                    {
+                        // a non-seekable stream should always be positioned at positionInStream + endPosition
+                        inputStream.ReadByte(); 
```

Comment:
This is the one thing in this file that feels odd to me.  From a quick read of this stream code, it looks like it should go through Fill when reading the input stream.  By doing the read here, I wonder if anything with respect to the input stream gets out of sync with the fields here.  I can't point to anything directly, and your code seems to pass the tests, so it hasn't broken seekable streams.  I would say that if this does break something, we can always fix it, but this is the thing that feels odd.

Reply:
My understanding is that seekable stream was allowed to be lazy, but I am preventing the non-seekable stream from being lazy. "Lazy" specifically referring to entering a state where positionInStream+endPosition != inputStream.Position and then later in Fill we do a Seek() to bring inputStream.Position back into agreement. 

Thinking on it I could probably make the non-seekable case match a little more closely by deferring the Read() until Fill(). I am planning to do a unit test and I'll try a change for this at the same time.